net: Automatically create hidden service, listen on Tor

laanwj opened the pull request:
Discussion in #6586.

Starting with Tor version 0.2.7.1 it is possible, through Tor's control socket API, to create and destroy 'ephemeral' hidden services programmatically.
https://stem.torproject.org/api/control.html#stem.control.Controller.create_ephemeral_hidden_service

This means that if Tor is running (and proper authorization is available), bitcoin automatically creates a hidden service to listen on, without user manual configuration. This will positively affect the number of available .onion nodes.
- When the node is started, connect to Tor through control socket
- Send `ADD_ONION` command
- First time:
  - Make it create a hidden service key
  - Save the key in the data directory for later usage
- Make it redirect port 8333 to the local port 8333 (or whatever port we're listening on).
- Keep control socket connection open for as long node is running. The hidden service will
  (by default) automatically go away when the connection is closed.

Adds command line options:
- `-listenonion` Automatically create Tor hidden service (default: 1)
- `-torcontrol=<ip>:<port>` "Tor control port to use if onion listening enabled (default: `127.0.0.1:9051`)

TODO:
- [x] SAFECOOKIE authentication support (see discussion in #6586)
- [x] HASHEDPASSWORD authentication: manually provide Tor credentials
- [x] Logging: most of the more specific logging should go into a debug category

laanwj committed:
net: Automatically create hidden service, listen on Tor

Starting with Tor version 0.2.7.1 it is possible, through Tor's control socket
API, to create and destroy 'ephemeral' hidden services programmatically.
https://stem.torproject.org/api/control.html#stem.control.Controller.create_ephemeral_hidden_service

This means that if Tor is running (and proper authorization is available),
bitcoin automatically creates a hidden service to listen on, without user
manual configuration. This will positively affect the number of available
.onion nodes.

- When the node is started, connect to Tor through control socket
- Send `ADD_ONION` command
- First time:
    - Make it create a hidden service key
    - Save the key in the data directory for later usage
- Make it redirect port 8333 to the local port 8333 (or whatever port we're listening on).
- Keep control socket connection open for as long node is running. The hidden service will
  (by default) automatically go away when the connection is closed.

petertodd committed:
Better error message if Tor version too old

laanwj committed:
torcontrol improvements and fixes

- Force AUTHCOOKIE size to be 32 bytes: This provides protection against
  an attack where a process pretends to be Tor and uses the cookie
  authentication method to nab arbitrary files such as the
  wallet
- torcontrol logging
- fix cookie auth
- add HASHEDPASSWORD auth, fix fd leak when fwrite() fails
- better error reporting when cookie file is not ok
- better init/shutdown flow
- stop advertizing service when disconnected from tor control port
- COOKIE->SAFECOOKIE auth

jgarzik commented:
I like the feature - concept ACK

Trying to think through whether there are edge cases that persuade us to default this 'off'  Leaning towards default-on as presented.

dcousens commented:
concept ACK

btcdrak commented:
concept ACK

sipa commented:
2015-09-26 00:03:11 [tor] Using COOKIE authentication, reading cookie authentication from /var/run/tor/control.authcookie
2015-09-26 00:03:11 [tor] Authentication cookie not found, is inaccessible, or is not exactly 64 bytes

I have ControlPort 9051 and CookieAuthentication in torrc.

pstratem commented:
Check the permissions
On Sep 25, 2015 5:07 PM, "Pieter Wuille" <notifications@github.com> wrote:

> 2015-09-26 00:03:11 [tor] Using COOKIE authentication, reading cookie
> authentication from /var/run/tor/control.authcookie
> 2015-09-26 00:03:11 [tor] Authentication cookie not found, is
> inaccessible, or is not exactly 64 bytes
>
> I have ControlPort 9051 and CookieAuthentication in torrc.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/6639#issuecomment-143380439>.
>

laanwj commented:
@sipa the check was broken, the cookie is 32 bytes not 64. That said, we should have a more specific error message.

theuni commented on src/torcontrol.cpp:
```diff
@@ -0,0 +1,580 @@
+    if (f == NULL)
+        return false;
+    if (fwrite(data.data(), 1, data.size(), f) != data.size())
+        return false;
```
false case needs fclose too

laanwj replied:
oops good catch

theuni commented:
Can the shutdown detection be changed to work like the httpserver? ie Init calls InterruptTorControl() which starts the immediate shutdown process?

theuni commented on src/torcontrol.cpp:
```diff
@@ -0,0 +1,580 @@
+
+bool TorControlConnection::Disconnect()
+{
+    if (b_conn)
```
Need to undo the AddLocal here?

theuni commented on src/torcontrol.cpp:
```diff
@@ -0,0 +1,580 @@
+    this->disconnected = disconnected;
+
+    // Finally, connect to target
+    if (bufferevent_socket_connect(b_conn, (struct sockaddr*)&connect_to_addr, connect_to_addrlen) < 0) {
```
Don't we want logic to retry this every n seconds if it fails? If I launch Bitcoin and Tor a minute later, I would expect that it would be detected and used after a few minutes.

laanwj replied:
That's what `RECONNECT_TIMEOUT_START` / `RECONNECT_TIMEOUT_EXP` is supposed to do. When I last checked, it was working.

theuni replied:
It looks like that only works on disconnection, though. If the first attempt fails on socket connection and the callbacks never start, I don't see how the reconnect logic kicks in. Am I misreading?

theuni replied:
Nevermind, I misremembered this function's return meaning. I thought it was possible to get an immediate failure in some cases, but a quick test shows that's not the case.

theuni commented:
(Not for this PR) Looks like it'd be pretty simple to add support for connecting via Unix socket (/var/run/tor/control) as well.

theuni commented:
Concept ACK. This looks like a great feature.

laanwj commented:
@cfields 

> Can the shutdown detection be changed to work like the httpserver? ie Init calls InterruptTorControl() which starts the immediate shutdown process?

Yes, that sounds better. I have not done this before because it means making the libevent instance thread-safe - `Interrupt` happens from a different thread. But as we need multi-threaded libevent anyway from the http server that's a bad reason.

> (Not for this PR) Looks like it'd be pretty simple to add support for connecting via Unix socket (/var/run/tor/control) as well.

Good idea.

laanwj commented:
Ok: addressed all of @theuni's comments (apart from the UNIX socket, which we can add later).
Still have SAFECOOKIE auth support on my TODO list above, it is superior to COOKIE (no potential exposure of data), and COOKIE will eventually be deprecated.
This does require use of HMAC-SHA256. Luckily we have an implementation of that for BIP32, so I think it's doable.

laanwj committed:
doc: Mention Tor listening in release notes

laanwj commented:
Also mentioned in release notes now.

petertodd commented:
ACK

Tests done:

- [X] Restart reuses cached onion private key
- [X] SAFECOOKIE auth
- [X] HASHEDPASSWORD auth
- [X] Onion advertisement stops with tor disconnects
- [X] Onion advertisement starts when tor restarted

sipa commented:
No code review, did a weak test (it correctly reports my Tor version is too old).

MarcoFalke commented:
That's great! I will try to look into this soon.

laanwj committed:
doc: update docs for Tor listening

- add new data directory files for 0.12 to doc/files.md
- mention torcontrol in doc/tor.md

petertodd commented:
utACK squashme commits

MarcoFalke commented on src/torcontrol.cpp:
```diff
@@ -0,0 +1,673 @@
+        service = CService(service_id+".onion", GetListenPort(), false);
+        LogPrintf("tor: Got service ID %s, advertizing service %s\n", service_id, service.ToString());
+        if (WriteBinaryFile(GetPrivateKeyFile(), private_key)) {
+            LogPrint("tor", "tor: Cached service private key to %s\n", GetPrivateKeyFile());
```
Nit: "Write cached ser.."?

MarcoFalke replied:
Also, mind to pass this file through

`contrib/devtools/clang-format.py clang-format src/torcontrol.cpp `?

MarcoFalke replied:
Nit: Please update https://github.com/bitcoin/bitcoin/blob/master/doc/files.md

MarcoFalke commented:
Somehow I can set up a `-onlynet=onion` node and connect to it (sometimes) but it never established outgoing connections:

```
2015-11-11 22:05:45 tor: ADD_ONION succesful
2015-11-11 22:05:45 tor: Got service ID moheur3skn7jbca4, advertizing service moheur3skn7jbca4.onion:18333
2015-11-11 22:05:45 tor: Cached service private key to /home/test/.bitcoin/testnet3/onion_private_key
2015-11-11 22:05:45 AddLocal(moheur3skn7jbca4.onion:18333,4)
2015-11-11 22:05:46 Cannot connect to t6xj6wilh4ytvcs7.onion:18333: unsupported network
2015-11-11 22:05:46 Cannot connect to 4azvkpo55qaskhjl.onion:18333: unsupported network
2015-11-11 22:05:47 Cannot connect to nkf5e6b7pl4jfd4a.onion:18333: unsupported network
2015-11-11 22:05:48 Cannot connect to 4zhkir2ofl7orfom.onion:18333: unsupported network
```

gits7r commented:
you are mistaking. it's -onlynet=tor . you confuse onion with -onion=127.0.0.1:9050 where you instruct bitcoin to use that socks5 proxy to connect to onion peers (from tor network). While -onlynet=tor will make sure you connect only to onion (Tor) peers, -onion=xxx will only provide a gateway for these peers but connect to other kind of peers also.

MarcoFalke commented:
Then, why does `-?` doesn't even mention `tor`?

```
  -onlynet=<net>
       Only connect to nodes in network <net> (ipv4, ipv6 or onion)
```

laanwj commented:
onlynet=onion should work. But is unrelated to this pull. Did you pass a proxy?

MarcoFalke commented:
I tried with and without proxy:
```bitcoin/src/bitcoind  -debug=tor  -onlynet=onion -proxy=127.0.0.1:9050 -torpassword=bar```

But maybe my VPS provider is blocking tor...

laanwj commented:
They could, but that whouldn't result in  errors like

    2015-11-11 22:05:46 Cannot connect to t6xj6wilh4ytvcs7.onion:18333: unsupported network

Instead you'd get proxy errors. It's not using the proxy, why is that (again, this is seemingly unrleated to this pull, maybe file a new issue)

MarcoFalke commented:
tested ACK 58ef0ff
